File: LeetCode problem set/101_Symmetric_Tree.cpp
#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int val) {
    this->val = val;
    left = NULL;
    right = NULL;
  }
};

class Solution {
private:
  bool check(TreeNode *left, TreeNode *right){
    if(!left || !right)
      return left == right;
    if(left->val == right->val) {
      return check(left->left, right->right) && check(left->right, right->left);
    }
    return false;
  };
public:
  bool isSymmetric(TreeNode* root) {
    if(!root) return true;
    return check(root->left, root->right);
  }
};

int main() {
  ios_base::sync_with_stdio(0), cin.tie(0), cin.tie(0);
  Solution s;
  TreeNode *root = new TreeNode(1); 
  root -> left = new TreeNode(3); 
  root -> right = new TreeNode(3); 
  root -> left -> left = new TreeNode(4); 
  root -> left -> right = new TreeNode(6);
  root -> right -> left = new TreeNode(6);
  root -> right -> right = new TreeNode(4);
  
  if (s.isSymmetric(root)){
    cout<<"The binary tree is symmetric."<<endl;
  }
  else{
    cout<<"The binary tree is asymmetric."<<endl;
  }
  return 0;
  return 0;
}

File: LeetCode problem set/101_Symmetric_Tree.js
class TreeNode {
  constructor(val, left, right) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

const isMirror = (left, right) => {
  if (!left || !right) return left == right;
  if (left.val === right.val)
    return isMirror(left.left, right.right) && isMirror(left.right, right.left);
};

const isSymmetric = (root) => {
  if (!root) return true;
  return isMirror(root.left, root.right) ? true : false;
};

const root = new TreeNode(1);
root.left = new TreeNode(2);
root.right = new TreeNode(2);
root.left.left = new TreeNode(3);
root.left.right = new TreeNode(5);
root.right.left = new TreeNode(4);
root.right.right = new TreeNode(3);

console.log(isSymmetric(root));
